### ISSUE TITLE
Create system for a centralized contrib repository (Composer, Github service, project.module, or ?)

### ISSUE THREAD

**Issue by jenlampton** [backdrop-ops/backdropcms.org#8]
In order to get Backdrop listed as a github service (so that contrib module/theme authors can register their github projects as Backdrop extensions) we'll need to write our own code for it.

Here's the repo:
https://github.com/github/github-services

Here's the documentation:
https://github.com/github/github-services/blob/master/CONTRIBUTING.md

**Comment by quicksketch** [comment]
I filed an issue similar to this one, but with a particular focus on adding version information to GitHub-hosted projects at https://github.com/backdrop-ops/backdropcms.org/issues/51.

**Comment by quicksketch** [comment]
It's been a while since we had updates here. For now we'll pursue the path of least resistance, which is simply porting Project module (with the sub-module "project_release.module"). We need the the functionality provided by project_release to power our update server. While we could rebuild an update server that pulled from a different repository, this is definitely the most direct approach to getting our update server running.

Project module includes *a lot* of unnecessary code for our purposes however. We don't need direct integration with VCS back-ends, issue management, maintainership, or particularly granular permissions. I expect we'll want to trim out a significant portion of the module in our port.

**Comment by quicksketch** [comment]
I'll be uploading my contrib version of Project module for Backdrop later this week.

**Comment by docwilmot** [comment]
:+1:

**Comment by quicksketch** [comment]
https://github.com/backdrop-contrib/project is up and is almost working at this point.

- Project: Fully working to the extent that you can create project nodes and the "project behavior" system is in place.
- Project Release: Simplified down significantly to reduce use of Fields and Taxonomy where it didn't make sense. There was a lot of hard-coded functionality that tied directly into specific fields and had dependencies on a few contrib modules that weren't really necessary. Releases had been stored as Drupal-owned files, but since we're hosting on GitHub, I switched these to simple URL fields instead.
- Project Usage: This isn't working quite yet but it's coming along. Apparently D7's usage systems is tied into both Varnish log files and MongoDB(?) It's a bit crazy.

Both project release and project usage are dependent upon Drush cron tasks, but they're so tied into the Drupal.org infrastructure that I think we'll be better off porting these scripts from D6, when things were MySQL-driven and simpler. They also were just plain PHP scripts instead of Drush, which will allow us to deploy this while we work out the drush integration.

We don't yet have GitHub integration in the port, right now it's pretty much what you would see in the D7 version of Project, but less complicated. I plan to add a "project_github.module" to the the repository to start providing the integration between the two systems.

**Comment by quicksketch** [comment]
I've added initial GitHub integration to https://github.com/backdrop-contrib/project in the new GitHub Project sub-module. I moved that packaging script (https://github.com/backdrop-ops/backdropcms.org/issues/51) into this module so that we can use a single web hook to do both the packaging and the automatic release creation.

I've also integrated the project nodes with the default branch README.md files, so that Backdrop can use the README.md as the project description instead of needing to manually update the description on BackdropCMS.org.

The only part left in the GitHub integration is to automatically create the project nodes themselves when a new project is created. After that, we need to get the update XML working and Project Usage module updated. But it's coming along! I can be installed, projects and releases created, and mostly synced with GitHub. We're having design meetings with @dariusgarza today to review the proposed designs for the project repository browser.

**Comment by sirkitree** [comment]
:octocat: :dancer:

**Comment by docwilmot** [comment]
:notes:  :+1:

**Comment by jenlampton** [comment]
Initial designs for this are looking fabulous, we're now moving on to designs for the module node itself.

**Comment by quicksketch** [comment]
> The only part left in the GitHub integration is to automatically create the project nodes themselves when a new project is created.

This is now done too. Only I found that we can only really automatically create project nodes at *release* time, not at the time of repository creation, as repositories usually will be empty (with maybe a README) at the time of creation on GitHub. It doesn't do us a lot of good to create the project on BackdropCMS.org with so little information.

> After that, we need to get the update XML working

This is working too! We've got update XML being generated. I can even point my local Backdrop install at itself to see the "available updates" that I've created as project releases locally:

![update-status](https://cloud.githubusercontent.com/assets/100206/7131883/dd7cfe14-e238-11e4-9320-bbb4ca848f46.png)

> and Project Usage module updated.

I've started working on this. Project Usage is comparatively a very simple module. I hope to have it working well later this week.

We're nearly at a point where we can start implementing on backdropcms.org. It might be a good idea to get it up and running, doing all the packaging, and start seeding the site with project and project release nodes while we hammer out the designs. Then we'll have content available on which we can begin theming, and it'll flesh out the project listing for when we take it live.

**Comment by klonos** [comment]
Great job!! Thanx Nate.

**Comment by biolithic** [comment]
Great job!! Thanx Nate. !!

**Comment by quicksketch** [comment]
> Where are modules/themes going to be listed? Are we to create a new "Extend" section in the main site?

Initially our designs call for adding Modules, Layouts, and Themes as top-level items in the left sidebar. We'll also kill the "Home" link at the time. I think there's a good chance we'll need to put these under a single top-level page, but since the number 1 reason I visit Drupal.org (other than issues) is to find and download modules, and I expect it will be this for Backdrop as well. So I'm not opposed to making it top-level until space becomes a problem.

**Comment by develCuy** [comment]
@quicksketch, you are following Drupal standards. How about doing something new and different? Community tried to give themes and modules a better position in d.o. Also, the usage stats are obscure, just a few people have access to them (I mean the full data, not "usage stats" page).

Also, Themes are a different world than Modules, they both deserve their own identity in terms of graphic design and functionality of respective section. For example, nice and big thumbnails of each theme in the list.

Well, there is a lot of ideas that come to my mind, by the root of everything is: Let's innovate a bit here, since there is room, time and energy for it in this particular moment.

My 2c.

**Comment by quicksketch** [comment]
@develCuy: So +1 for top-level links?

> Also, the usage stats are obscure, just a few people have access to them (I mean the full data, not "usage stats" page).

I think usage stats are totally public, but perhaps not well-linked on drupal.org. From the looks of project.module, the full history of stats is actually discarded after a time-period, so my guess is that any missing stats have actually been removed from the DB.

It might be neat to give usage stats a more prominent location, but initially they won't be very exciting until we've collected stats for at least a few months. But totally, I see what you're saying. We don't have to do the same thing. If anything, I'd like to see our minimalist approach to content used to our advantage to make the more important content easier to find.

**Comment by quicksketch** [comment]
Oh yeah! Our first auto-generated project nodes and releases are up on backdropcms.org!

Project module: https://backdropcms.org/node/108
Form block module: https://backdropcms.org/node/110

Apparently we don't have Pathauto on BackdropCMS.org. We can add that shortly or wait for 1.1.0.

Unfortunately, auto-generating projects requires **a minor change to .info files**. Because our contributed project group is a mishmash of themes, modules, and layouts, we have no way of identifying which is which. Given a module and theme .info file, there's no indication what type of project it is. Layouts can be inferred by the presence of "regions" in the .info file, but that's it.

So... in order to properly create a node on BackdropCMS.org, projects *must* specify a `type` in their .info files like such:

```
name = Project
description = Base module for identifying nodes as projects (or relating to projects).
type = module
package = Project
backdrop = 1.x
```

Then when BackdropCMS.org gets the initial ping to create a matching node, it knows which type of node to create (since themes and modules have different fields in their node types).

What I'd like to do now is get a bunch of projects creating initial releases (and including these new .info file lines) to start populating BackdropCMS.org with contrib projects. And now that the technical requirements are met, we can start implementing the designs from @dariusgarza.

**Comment by docwilmot** [comment]
:clap: 
Great news.
Will 'release' all my ports shortly.

**Comment by sirkitree** [comment]
Release nodes should have a link back to the github project.

**Comment by docwilmot** [comment]
This also doesnt seem to have tags or author data. A "See more themes by Darius", or "List Media modules", or "Filter by fullwidth layouts" would be impossible then?

**Comment by quicksketch** [comment]
Hi guys, great feedback.

> Release nodes should have a link back to the github project.

I thought this as well while doing the port. Then I realized that the breadcrumb contained a link back to the project. But we don't have breadcrumbs on BackdropCMS.org, doh! I thought the link was a good idea though, I'll put one in.

> This also doesnt seem to have tags or author data. A "See more themes by Darius", or "List Media modules", or "Filter by fullwidth layouts" would be impossible then?

The only information we have right now is the data we can get from GitHub. Since GitHub doesn't have any kind of categorization, we can only work with information we can pull from the project itself. We'll start having on-site registration in the near future, allowing users to customize project nodes more extensively, but I'd love to get as much data as possible from GitHub.

This made me realize that we should automate the creation of the thumbnail image for themes. These already include a "preview.png", which we should pull in and attach to theme projects.

We could also pull in categories from the .info file (or at least the current "Group" listing) but I think we'd want to curate that list of categories on the site.

Author data can also be pulled in from GitHub username, but we don't have information about that user on BackdropCMS.org. Maybe we should bind GitHub usernames to BackdropCMS.org UIDs so we could directly make this connection?

**Comment by docwilmot** [comment]
>The only information we have right now is the data we can get from GitHub

Maybe start [adding more metadata in .info files](https://github.com/backdrop/backdrop-issues/issues/119) and also [adding tags to info files](https://github.com/backdrop/backdrop-issues/issues/37)?

> Author data can also be pulled in from GitHub username
> Maybe we should bind GitHub usernames to BackdropCMS.org UIDs

But we dont have any way to know who's maintaining a repo, except parsing the README, which seems a bit not so nice.

Is there a plan to be able to add a project by directly creating a node on BD.org? Or will it only via this automatic release-reading process?

**Comment by quicksketch** [comment]
> But we dont have any way to know who's maintaining a repo, except parsing the README, which seems a bit not so nice.

Yeah, the only thing we'll be able to do accurately is assign the owner of the node, assuming we start tracking GitHub usernames on backdropcms.org.

> Is there a plan to be able to add a project by directly creating a node on BD.org?

Yes, as soon as we start allowing logins, you should be able to create/edit projects (and releases) directly on the site as well. That might be a temporary solution for distributed repositories instead of all needing to be in the backdrop-contrib group.

**Comment by klonos** [comment]
> So +1 for top-level links?

Yes, :+1: form me.

> Oh yeah! Our first auto-generated project nodes and releases are up on backdropcms.org!

:tada: :balloon: :clap: 

Also :+1: for linking back to the github projects and author (maintainer) data. What I would also like to see implemented in BackdropCMS.org is some issues like these:

#59

> Apparently we don't have Pathauto on BackdropCMS.org. We can add that shortly or wait for 1.1.0.

Add it now and test the upgrade path to 1.1.0 when it's out. We could use the SEO boost from having the actual project names in the URL :wink:

**Comment by docwilmot** [comment]
What about adding data to `.info` files?

```
tags[] = media
tags[] = fields
tags[] = user-interface
```
and
```
maintainers[] = @quicksketch
maintainers[] = @klonos
```
together with from [this issue](https://github.com/backdrop/backdrop-issues/issues/119):

```
source = url
support = url
documentation = url
```

Any technical reasons why not (apart from someone has to write the code to make this useful onB.org of course). I'm considering the 'inconvenience' of uploading to GitHub and still then having to go and edit a node on B.org. If it can all be done automagically, seems better no?

**Comment by kreynen** [comment]
+1 for any process that doesn't require logging into to multiple sites

**Comment by docwilmot** [comment]
How do updates work? If I edit the info node on B.org, does that update the README on GitHUb, or vice versa?

I'm asking mainly because I was holding off on creating releases until a response to the idea I proposed above, as I would want to have to release them all over again. 

But if updates would be picked up, I may as well go on.

**Comment by quicksketch** [comment]
> If I edit the info node on B.org, does that update the README on GitHUb, or vice versa?

Editing the description is locked on B.org if the README is being pulled from GitHub. Any updates to the README.md on GitHub are picked up immediately by B.org upon pushing them to GitHub. The option to edit the node on B.org is intended to be used only if you want an entirely different description from your README.md, in which case syncing becomes disabled. Here's a screenshot from the node edit form:

![project-edit](https://cloud.githubusercontent.com/assets/100206/7355864/7fd3c076-ecd9-11e4-95d7-e848d7bbd986.png)

So in short, there's no reason to delay. All your changes to the README will be picked up immediately.

**Comment by docwilmot** [comment]
Thanks, but I actually was actually intending to ask about changes to the `.info` file, not the README, sorry. But I suspect that must be similar?

**Comment by quicksketch** [comment]
Ah! Yeah well the only thing we pull from the .info file at this point is the name of the project. All of these options for pulling categories, etc. hasn't been implemented yet.

**Comment by quicksketch** [comment]
Just a note here that I've closed https://github.com/backdrop-ops/backdropcms.org/issues/22 (provide an update server). As noted in the [1.1.0 Release mailer](http://us7.campaign-archive2.com/?u=8a016d204af55747c1fef5769&id=846c1e27d9), our update server is now enabled!

We've now got 45 projects listed on BackdropCMS.org. We just need to get the views created for the project listing and start styling them.
